### DISCUSSION TITLE
[UX] keyboard shortcuts change suggestions

### DISCUSSION THREAD

**Discussion by adriangalilea** [darrenburns/elia#31] [Ideas]
Hello,

Very pleased with v1.0 thanks for your work @darrenburns,

(re: https://github.com/darrenburns/elia/pull/23#issuecomment-2094990166)

I would like to propose some UX changes/additions on shortcuts and navigation, most of them are minor and a matter of taste.

## <kbd>shift</kbd>+<kbd>enter</kbd> send message when on input text mode
current: <kbd>ctrl</kbd>+<kbd>j</kbd> send message
I think it's fine to have a shortcut that sends the message without focus.

But it should also have <kbd>shift</kbd>+<kbd>enter</kbd> to send a message when focused in the input box. We all have muscle memory for this.

Why `j` tho? (I'm assuming not moving the finger from resting position? that's fair)

## <kbd>q</kbd> exit mode
I would like to have exit mode just like lazygit, <kbd>q</kbd>, I understand this clashes with the focus mode that defaults to an input box, I would prefer to have <kbd>q</kbd> rather than defaulting to input text mode.

The focus could be on the pane and <kbd>enter</kbd> to actually input text, so now you can have q to exit without interfering with the text input.

## navigation
default screen let's you press <kbd>↓</kbd> to change pane, but then you can not press <kbd>↑</kbd> to go back, this feels unnatural, so maybe, like lazygit again, you could <kbd>←</kbd> <kbd>→</kbd> to navigate between panes, and <kbd>↓</kbd><kbd>↑</kbd> to navigate between elements of a pane.

## <kbd>enter</kbd> to select text mode
Keeping navigation consistent if the above suggestion sticks, I would do <kbd>enter</kbd> to select instead of <kbd>space</kbd> then:
<kbd>esc</kbd> to exit without copying.
<kbd>space</kbd>/<kbd>enter</kbd> to copy and exit the text selection mode

Also, I would like to have <kbd>v</kbd> helix style selection mode to avoid having to sustain <kbd>shift</kbd> depressed for the selection.

## <kbd>f1</kbd> <kbd>f2</kbd> clashes with terminal shortcuts I have
Perhaps we can select on the config how some shortcuts behave, instead of trying to make some that try and fail to work for everyone.

I would prefer to have <kbd>?</kbd> for help, this only works of course if you like the not defaulting to input text.

**Comment by darrenburns** [comment]
Thanks for your feedback!

I'm pretty sure shift+enter doesn't work on most terminals sadly.

For the exit stuff it's a bit of a trade-off: I think the most common user flow would be: open app, enter message. I'd like to keep the input focused by default to reduce that friction. escape key might work though.

Agree on the navigation. It just never made it into 1.0, but it will come.

The reason I didn't have enter for select mode was because previously this would open a "message details" panel. I think this suggestion makes sense though. Also the idea about presses esc to exit selection mode is needed - I actually keep pressing it myself!

I like the idea of toggle selection!

I'm not 100% sure how we'd have bindings pulled from config quite yet. Will need to do some research on the Textual side.

**Reply by adriangalilea** [reply]
I'm glad you liked the suggestions,

> I'm pretty sure shift+enter doesn't work on most terminals sadly.

Oh, that's sad, it would be great to have it on for whatever terminals that do support it, my muscle memory seems to refuse  <kbd>ctrl</kbd>+<kbd>j</kbd> :)

> For the exit stuff it's a bit of a trade-off: I think the most common user flow would be: open app, enter message. I'd like to keep the input focused by default to reduce that friction. escape key might work though.

I totally understand this,  and as you mentioned you could start with the input text by default but still have esc to exit input mode and have the focus on the pane, that way implementing <kbd>q</kbd> is feasible and it remains consistent with lazygit and some other TUI's I'm using(taskwarrior-tui for instance), it adds mental overhead to have several TUI's whose exit keybinding are different :)

> I'm not 100% sure how we'd have bindings pulled from config quite yet. Will need to do some research on the Textual side.

Got it, I'm asking because I had <kbd>F1</kbd> set up for kitty, and went ahead and changed it to <kbd>F2</kbd> on kitty just to realise in the chat mode that <kbd>F2</kbd> was also used 😆 so it may be convenient to others as well.

**Reply by darrenburns** [reply]
@adriangalilea Often you can remap keybindings at the terminal emulator level for more control. e.g. you could set shift+enter to send the same escape sequence as ctrl+j :)

Would you expect the `q` keybinding to work anywhere in Elia, as long as a text area isn't currently focused?

**Reply by adriangalilea** [reply]
> @adriangalilea Often you can remap keybindings at the terminal emulator level for more control. e.g. you could set shift+enter to send the same escape sequence as ctrl+j :)

Will check thanks

> Would you expect the q keybinding to work anywhere in Elia, as long as a text area isn't currently focused?

I see 2 options:

<kbd>q</kbd> exiting the program anywhere that's not inputting text.
<kbd>esc</kbd> working in "levels" pretty much as it does now but if you are on the initial position "Enter your message pane..." and the input text mode is  already escaped the next <kbd>esc</kbd> quits the app. This way the flow is consistent and predictable.

Both of these are compatible, up to you ;)

**Comment by darrenburns** [comment]
@adriangalilea Some of the changes from this thread are now implemented and available in the latest version.

There's now visual select mode (press `v` - I still want to make the UI clearer for this though), the bindings are a little more intuitive (I think, but still got a little work to do on the up/down arrows), and you can press `q` to quit. `esc` now moves focus "closer to home", which is harder to explain precisely, but it feels better for me.

If you try it out, let me know how it goes!
